### PR TITLE
closes #93

### LINKED ISSUE #93: Move backbone model initialization to backbones.py
### Description:
Currently, the initialization of all backbone models, such as ResNet 50, 101, ViT, and EffNet, is hardcoded in the models/base.py file. This approach lacks modularity and can lead to code duplication. To improve the code structure and maintainability, it is proposed to refactor the code by moving the backbone model initialization logic to a dedicated folder called 'backbones'. Each backbone model will have its own file, for example, resnets.py. Additionally, a build_backbone function will be created to handle the initialization of the backbones based on the provided configuration.

### Motivation:
The current implementation suffers from several drawbacks. Firstly, having all backbone model initializations hardcoded in a single file makes it challenging to locate and modify specific backbone configurations. Secondly, it leads to code duplication if multiple files require the same backbone model. This lack of modularity can hinder the scalability and maintainability of the codebase.

### Proposed Solution:

1. Create a new folder called 'backbones' in the project directory.
2. Inside the 'backbones' folder, create a separate Python file for each backbone architecture model, such as resnets.py, vits.py, etc.
3. Move the corresponding backbone model initialization code from the base.py file to their respective backbone files in the 'backbones' folder.
4. Define a function called build_backbone in the base.py file, which takes the configuration as input and returns the initialized backbone model.
5. In the build_backbone function, based on the provided configuration, dynamically import the appropriate backbone file from the 'backbones' folder and initialize the corresponding backbone model.
6. Update the relevant parts of the codebase to call the build_backbone function instead of the hardcoded backbone initializations.

### Benefits:

- Improved modularity: By separating each backbone model into its own file, it becomes easier to locate and modify specific backbone configurations without affecting unrelated code.
- Code reuse: The separation of backbone models into individual files eliminates code duplication, enabling multiple files to utilize the same backbone model initialization.
- Maintainability: The proposed folder structure and build_backbone function enhance the overall maintainability of the codebase, making it easier to add, modify, or remove backbone models in the future.
- Readability: The new structure improves code readability by isolating backbone model logic into dedicated files, enhancing code comprehension, and reducing cognitive load.

Please let me know if you need any further clarification or have any questions regarding the proposed refactor.

### PR DESCRIPTION
**Pull request by themattinthehatt**
refactor backbone code and integrate context/unsupervised loss features into models with transformer backbones

### CODE REVIEW

**Comment by themattinthehatt** [comment]
@luca-medeiros take a look and let me know what you think

File: lightning_pose/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import Literal
```
Comment (danbider) [inline lightning_pose/models/__init__.py:1]: very good idea to centralize

File: lightning_pose/models/backbones/torchvision.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
+import torch
+import torchvision.models as tvmodels
+from typeguard import typechecked
+
+
+@typechecked
+def build_backbone(
+    backbone_arch: str, pretrained: bool = True, model_type: str = "heatmap", **kwargs,
```
Comment (danbider) [inline lightning_pose/models/backbones/torchvision.py:8]: the `pretrained` argument is deprecated for `torchvision.models`? we get tons of warnings.
we should use `weights` instead. checkout the new neat guidelines here; it'll enable loading different pretrained weights for the same resnet50 archi.
https://pytorch.org/vision/stable/models.html

File: lightning_pose/models/base.py [hunk 3]
```diff
@@ -3,17 +3,11 @@
 from omegaconf import DictConfig
```
Comment (danbider) [inline lightning_pose/models/base.py:3]: looks cleaner!

File: lightning_pose/models/base.py [hunk 4]
```diff
@@ -22,37 +16,12 @@
     SemiSupervisedBatchDict,
     SemiSupervisedHeatmapBatchDict,
 )
-
+from lightning_pose.models import ALLOWED_BACKBONES
```
Comment (danbider) [inline lightning_pose/models/base.py:19]: nice

File: lightning_pose/models/heatmap_tracker.py [hunk 5]
```diff
@@ -142,9 +127,10 @@ def __init__(
         self.rmse_loss = RegressionRMSELoss()
 
         # necessary so we don't have to pass in model arguments when loading
-        # added loss_factory_unsupervised which might come from the SemiSupervisedHeatmapTracker.__super__(). Otherwise it's ignored.
+        # added loss_factory_unsupervised which might come from the
+        # SemiSupervisedHeatmapTracker.__super__(). Otherwise it's ignored.
         # that's important so that it doesn't try to pickle the dali loaders.
-        self.save_hyperparameters(ignore=["loss_factory", "loss_factory_unsupervised"])  # cannot be pickled
```
Comment (danbider) [inline lightning_pose/models/heatmap_tracker.py:147]: i think we should keep the "cannot be pickled" because we can run into issues if we don't do that

File: tests/models/test_base.py [hunk 6]
```diff
@@ -9,121 +10,199 @@
...
 _TORCH_DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
 
-BATCH_SIZE = 4
+BATCH_SIZE = 2
 HEIGHTS = [128, 256, 384]  # standard numbers, not going to bigger images due to memory
 WIDTHS = [120, 246, 380]  # similar but not square
-BACKBONES = ["resnet18", "resnet34", "resnet50", "resnet101", "resnet152"]
+RESNET_BACKBONES = ["resnet18", "resnet34", "resnet50", "resnet101", "resnet152"]
+EFFICIENTNET_BACKBONES = ["efficientnet_b0", "efficientnet_b1", "efficientnet_b2"]
+VIT_BACKBONES = ["vit_b_sam"]  # "vit_h_sam" very large (2.6GB), takes too long to download/load
```
Comment (danbider) [inline tests/models/test_base.py:18]: `vit_h_sam` is expected to have 2.6GB and long downloading. that's how it is with big models and depends on internet connection.
If we were to commit to transformers (sam or not) we might want to move to float16 represntation instead of float32. this will eliminate memory by half and will allow for bigger nets and batchsizes.

File: tests/models/test_base.py [hunk 7]
```diff
@@ -9,121 +10,199 @@
...
-                backbone=backbone, last_resnet_layer_to_get=-2
+                backbone=backbone, image_size=HEIGHTS[idx_image]
             ).to(_TORCH_DEVICE)
+            fake_image_batch = torch.rand(
+                size=(BATCH_SIZE, 3, HEIGHTS[idx_image], HEIGHTS[idx_image]),
+                device=_TORCH_DEVICE,
+            )
             representations = model(fake_image_batch)
-            assert representations.shape == repres_shape_list_all_but_fc[ind]
+            assert representations.shape == shape_list_pre_pool[idx_image][idx_backbone]
```
Comment (danbider) [inline tests/models/test_base.py:202]: maybe add a comment? this is like the above but just to say that the rep dim depends on image size?

**Review by danbider — COMMENTED** [review]
looks good to me, I think we can merge after incorporating the `pretrained` --> `weights` change? also might be good to look at `torchvision.models` before closing this PR, to make sure it'll be easy to work with their recent syntax/models.

File: lightning_pose/models/backbones/torchvision.py (hunk 2 above)
Comment (themattinthehatt) [reply lightning_pose/models/backbones/torchvision.py]: thanks for pointing this out, updated now

**Review by luca-medeiros — APPROVED** [review]
Cool stuff! looking very neat Matt LGTM
